torch_geometric/loader/link_neighbour_loader.py

```diff
@@ -0,0 +1,95 @@
+from typing import Any, Callable, List, Tuple
+
+import torch
+from torch import Tensor
+
+from torch_geometric.data import Data
+from torch_geometric.loader.base import BaseDataLoader
+from torch_geometric.loader.neighbor_loader import NeighborSampler
+from torch_geometric.loader.utils import filter_data
+from torch_geometric.typing import InputEdges, NumNeighbors, OptTensor
+from torch_geometric.utils.mask import mask_to_index
+
+
+class LinkNeighborSampler(NeighborSampler):
+    def __init__(
+        self,
+        data: Data,
+        num_neighbors: NumNeighbors,
+        replace: bool = False,
+        directed: bool = True,
+    ):
+        super().__init__(data, num_neighbors, replace, directed)
+        self.data = data
+
+    def __call__(self, index: List[Tensor]):
+
+        batch_size = len(index)
+
+        # take start and end node from each edge then deduplicate
+        row, col = self.data.edge_index
+        node_index = torch.cat([row[index], col[index]], dim=0)
+        node_index = torch.unique(node_index)
+
+        # get sampled graph
+        node, row, col, edge, _ = super().__call__(index)
+        return node, row, col, edge, batch_size, index
+
+
+class LinkNeighborLoader(BaseDataLoader):
+    def __init__(
+        self,
+        data: Data,
+        num_neighbors: NumNeighbors,
+        input_edge_idx: InputEdges = None,
+        input_edge_labels: OptTensor = None,
+        replace: bool = False,
+        directed: bool = True,
+        transform: Callable = None,
+        **kwargs,
+    ):
+
+        # Remove for PyTorch Lightning:
+        if 'dataset' in kwargs:
+            del kwargs['dataset']
+        if 'collate_fn' in kwargs:
+            del kwargs['collate_fn']
+
+        # Save for PyTorch Lightning:
+        self.data = data
+        self.num_neighbors = num_neighbors
+        self.input_edge_idx = input_edge_idx
+        self.input_edge_labels = input_edge_labels
+        self.replace = replace
+        self.directed = directed
+        self.transform = transform
+        self.neighbor_sampler = LinkNeighborSampler(data, num_neighbors,
+                                                    replace, directed)
+
+        return super().__init__(self.get_input_edge_idx(),
+                                collate_fn=self.neighbor_sampler, **kwargs)
+
+    def transform_fn(self, out: Any) -> Tuple[Data, torch.Tensor]:
+        node, row, col, edge, batch_size, index = out
+        data = filter_data(self.data, node, row, col, edge,
+                           self.neighbor_sampler.perm)
+
+        data.batch_size = batch_size
+        data = data if self.transform is None else self.transform(data)
+
+        labels = self.get_input_edge_labels()[index]
```

Comment:
Shall we write this directly into `data`? (`edge_label` and `edge_label_index`)?

Reply:
Yeah I was wondering about that. I thought it might be a little bit confusing to have a property that didn't exist on all edges, but its definitely much cleaner.

Reply:
I think this replicates our `RandomLinkSplit`quite well in which we have `data.edge_index` for message passing` and `data.edge_label_index` for supervision (with potentially different shape).

Reply:
I've updated this to store `sampled_edges` and `sampled_edge_labels` (and `sampled_index`, maybe we can drop this one), so that it's easy to work with labels of edges that don't exist in the original graph.